File: 5월 1주차/leetcode-938-Range Sum of BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int mrounding(TreeNode* node, int low, int high, int sum) {
        if (node == nullptr) return sum;

        int lsum = sum;
        if (node->val > low) lsum = mrounding(node->left, low, high, sum);

        if (node->val >= low && node->val <= high) lsum += node->val;

        if (node->val >= high) return lsum;
        return mrounding(node->right, low, high, lsum);
    }
    int rangeSumBST(TreeNode* root, int low, int high) {
        return mrounding(root, low, high, 0);
    }
};

File: maureen72/Tree/Leetcode-938-RangeSumofBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def rangeSumBST(self, root, low, high):
        """
        :type root: TreeNode
        :type low: int
        :type high: int
        :rtype: int
        """
        stack = [root]
        sum = 0

        while stack:
            node = stack.pop()
            if node:
                if node.val > low:
                    stack.append(node.left)
                if node.val < high:
                    stack.append(node.right)
                if low <= node.val <= high:
                    sum += node.val
        return sum